File: SpringBootJpa/src/test/java/com/example/OneToManyTests.java
package com.example;

import com.example.onetomany.domain.Order;
import com.example.onetomany.repository.OrderRepository;
import com.example.onetomany.domain.Product;
import com.example.onetomany.repository.ProductRepository;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.transaction.annotation.Transactional;

@RunWith(SpringRunner.class)
@DataJpaTest
@Transactional
public class OneToManyTests {

    @Autowired TestEntityManager testEntityManager;
    @Autowired OrderRepository orderRepository;
    @Autowired ProductRepository productRepository;

    @Before
    public void before_init() {
        Product javaProduct = new Product("java", "java content");
        Product kotlinProduct = new Product("kotlin", "kotlin content");
        Product scalaProduct = new Product("scala", "scala content");

        productRepository.save(javaProduct);
        productRepository.save(kotlinProduct);
        productRepository.save(scalaProduct);

        Order order1 = orderRepository.save(new Order(3, ""));
        Order order2 = orderRepository.save(new Order(2, ""));
        Order order3 = orderRepository.save(new Order(1, ""));
        Order order4 = orderRepository.save(new Order(3, ""));
        Order order5 = orderRepository.save(new Order(2, ""));

        javaProduct.getOrders().add(order1);
        kotlinProduct.getOrders().add(order2);
        scalaProduct.getOrders().add(order3);
        javaProduct.getOrders().add(order4);
        kotlinProduct.getOrders().add(order5);

        testEntityManager.flush();
        testEntityManager.clear();
    }

    @Test
    public void test_productFindOne() {
        System.out.println(productRepository.findOne(1L));
    }

    @Test
    public void test_orderFindOne() {
        System.out.println(orderRepository.findOne(1L));
    }
}

File: SpringBootJpa/src/test/kotlin/com/example/kotlin/OneToManyTests.kt
package com.example.kotlin

import com.example.kotlin.onetomany.domain.Order
import com.example.kotlin.onetomany.domain.Product
import com.example.kotlin.onetomany.repository.OrderRepository
import com.example.kotlin.onetomany.repository.ProductRepository
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.test.context.junit4.SpringRunner
import java.util.*
import java.util.function.Consumer

@RunWith(SpringRunner::class)
@DataJpaTest
open class OneToManyTests {

    @Autowired
    lateinit var orderRepository: OrderRepository

    @Autowired
    lateinit var productRepository: ProductRepository

    @Test
    fun test_productFindByName() {
        // given
        insertBaseData()

        // when
        val product = productRepository.findByName(PRODUCT_NAME)

        // then
        assertThat(product.name).isEqualTo(PRODUCT_NAME)
    }

    private fun insertBaseData() {
        val javaProduct = productRepository.save(Product(name = "java", content = "java content"))
        val kotlinProduct = productRepository.save(Product(name = "kotlin", content = "kotlin content"))
        val scalaProduct = productRepository.save(Product(name = "scala", content = "scala content"))

        val javaOrders = Arrays.asList(Order(productCount = 3, product = javaProduct), Order(productCount = 3, product = javaProduct))
        val kotlinOrders = Arrays.asList(Order(productCount = 2, product = kotlinProduct), Order(productCount = 2, product = kotlinProduct))
        val scalaOrders = Arrays.asList(Order(productCount = 1, product = scalaProduct))

        orderRepository.saveAll(javaOrders)
        orderRepository.saveAll(kotlinOrders)
        orderRepository.saveAll(scalaOrders)

        println("========================= Product =========================")
        productRepository.findAll().forEach(Consumer<Product> { println(it) })
        println("========================== Order ==========================")
        orderRepository.findAll().forEach(Consumer<Order> { println(it) })
    }

    companion object {
        const val PRODUCT_NAME = "java"
    }
}
